Remove raw SQL queries

sevein opened the pull request:
Putting the agent on a debt detox!

sevein committed:
Remove raw SQL queries

jraddaoui commented:
Thanks @sevein! We were just talking about this yesterday, I'll take a look later today. I guess you forgot to remove the draft status from the PR :wink:.

sevein committed:
Update: remove explit In(time.UTC) call

Because the MySQL connector already forces UTC, the returned time.Time values
are already in UTC so calling In(time.UTC) add no effect.

sevein committed:
Update: preserve nil vs empty semantics

sevein committed:
Update: add datatypes/doc.go

sevein committed:
Update: clean up ingest_test.go

sevein committed:
Update: clean up workflow_test.go

sevein committed:
Update: use nullTime

sevein committed:
Update: clear fields

sevein committed:
Update: clean up workflow_test.go

jraddaoui commented on internal/ingest/convert.go:
```diff
@@ -69,7 +78,7 @@ func taskToGoa(task *datatypes.Task) *goaingest.SIPTask {
 		// TODO: Make Goa StartedAt a pointer to a string to avoid having to
 		// convert a null time to an empty (zero value) string.
-		StartedAt: ref.DerefZero(db.FormatOptionalTime(task.CompletedAt)),
+		StartedAt: ref.DerefZero(db.FormatOptionalTime(task.StartedAt)),
```
:man_facepalming: I recall seeing this issue in the UI, but we changed how the task are displayed and completely forgot about it, thanks!

jraddaoui commented on internal/ingest/goa.go:
```diff
@@ -177,56 +175,22 @@ func (svc *ingestImpl) ListSipWorkflows(
+	if _, err := svc.perSvc.ReadSIP(ctx, sipUUID); err == persistence.ErrNotFound {
 		return nil, &goaingest.SIPNotFound{UUID: payload.UUID, Message: "SIP not found"}
 	} else if err != nil {
 		return nil, goaingest.MakeNotAvailable(errors.New("cannot perform operation"))
```
Not for this PR, but we are starting to return internal errors (500) from these cases.

jraddaoui commented on internal/ingest/ingest.go:
```diff
@@ -146,14 +143,12 @@ func (svc *ingestImpl) UpdateSIP(
 }
 
 func (svc *ingestImpl) SetStatus(ctx context.Context, id uuid.UUID, status enums.SIPStatus) error {
```
Also not for this PR, but we should rename to `SetSIPStatus`. But we also mentioned the possibility to just have a general `UpdateSIP` method/event.

jraddaoui commented on internal/ingest/ingest.go:
```diff
@@ -163,19 +158,18 @@ func (svc *ingestImpl) SetStatus(ctx context.Context, id uuid.UUID, status enums
+	_, err := svc.perSvc.UpdateSIP(ctx, id, func(s *datatypes.SIP) (*datatypes.SIP, error) {
+		s.Status = enums.SIPStatusProcessing
+		if !startedAt.IsZero() {
+			s.StartedAt = sql.NullTime{
+				Time:  startedAt,
+				Valid: true,
+			}
```
Something else we discussed the other day, now without raw SQL, we plan to move to `time.Time` all over.

sevein replied:
That makes a lot of sense!

jraddaoui commented on internal/ingest/ingest.go:
```diff
@@ -163,19 +158,18 @@ func (svc *ingestImpl) SetStatus(ctx context.Context, id uuid.UUID, status enums
+		return s, nil
+	})
+	if err != nil {
+		return fmt.Errorf("error updating SIP: %v", err)
```
Another thing not for this PR, we should stop using "error" in the error messages.

jraddaoui commented on internal/persistence/ent/client/convert.go:
```diff
@@ -114,3 +101,42 @@ func convertBatch(batch *db.Batch) *datatypes.Batch {
 }
+
+// convertWorkflow converts an entgo `db.Workflow` representation
+// to a `datatypes.Task` representation.
```
... `datatypes.Workflow` representation.

jraddaoui commented on internal/persistence/ent/client/convert.go:
```diff
@@ -114,3 +101,42 @@ func convertBatch(batch *db.Batch) *datatypes.Batch {
+		UUID:        dbw.UUID,
+		TemporalID:  dbw.TemporalID,
+		Type:        dbw.Type,
+		Status:      enums.WorkflowStatus(uint(dbw.Status)), // #nosec G115 -- constrained value.
```
Another mental note not for this PR, move this to a proper enum like we did with the SIP statuses.

jraddaoui commented on internal/persistence/ent/client/convert.go:
```diff
@@ -114,3 +101,42 @@ func convertBatch(batch *db.Batch) *datatypes.Batch {
+
+	// Only populate Tasks if they were loaded, preserving nil vs empty semantics.
+	if dbw.Edges.Tasks != nil {
+		w.Tasks = make([]*datatypes.Task, 0, len(dbw.Edges.Tasks))
```
Nitpicking, not a big deal, but knowing the length of tasks there is a small improvement using `make([]*datatypes.Task, len(dbw.Edges.Tasks))` and using the index instead of append.

sevein replied:
Thanks, this shows I'm a bit rusty, the pre-sized slice trick is easy to forget!

jraddaoui commented on internal/persistence/ent/client/convert.go:
```diff
@@ -114,3 +101,42 @@ func convertBatch(batch *db.Batch) *datatypes.Batch {
+	return w
+}
+
+func nullTime(t time.Time) sql.NullTime {
+	if t.IsZero() {
+		return sql.NullTime{}
+	}
+	return sql.NullTime{Time: t.UTC(), Valid: true}
+}
```
Nice! until we get rid of `sql.NullTime`.

jraddaoui commented on internal/persistence/ent/client/workflow.go:
```diff
@@ -66,3 +69,96 @@ func (c *client) CreateWorkflow(ctx context.Context, w *datatypes.Workflow) erro
+	dbw, err := tx.Workflow.Query().
+		Where(workflow.ID(id)).
+		WithSip().
+		WithTasks().
```
I think we don't want the tasks on update, there may be a big set of them. At least in the WS event and in the dashboard we don't care about them:

https://github.com/artefactual-sdps/enduro/blob/main/dashboard/src/monitor-ingest.ts#L70-L73

jraddaoui commented on internal/persistence/ent/client/workflow.go:
```diff
@@ -66,3 +69,96 @@ func (c *client) CreateWorkflow(ctx context.Context, w *datatypes.Workflow) erro
+	if up.Status.IsValid() {
+		q.SetStatus(int8(up.Status)) // #nosec G115 -- constrained value.
+	}
+	if up.StartedAt.Valid {
+		q.SetStartedAt(up.StartedAt.Time.UTC())
+	} else {
+		q.ClearStartedAt()
+	}
+	if up.CompletedAt.Valid {
+		q.SetCompletedAt(up.CompletedAt.Time.UTC())
+	} else {
+		q.ClearCompletedAt()
+	}
```
Having conflicts about clearing, I guess it could be good for restarts, if we use the same workflow record.

jraddaoui commented on internal/persistence/ent/client/workflow.go:
```diff
@@ -66,3 +69,96 @@ func (c *client) CreateWorkflow(ctx context.Context, w *datatypes.Workflow) erro
+		Only(ctx)
+	if err != nil {
+		return nil, rollback(tx, newDBError(err))
+	}
```
We don't need to read the workflow again, use the one returned by `Save`. We don't care about the tasks, but we care about the SIP UUID. Check the storage version:

https://github.com/artefactual-sdps/enduro/blob/main/internal/storage/persistence/ent/client/workflow.go#L67-L113

sevein replied:
Nice, good catch!

jraddaoui commented on internal/persistence/ent/client/workflow.go:
```diff
@@ -66,3 +69,96 @@ func (c *client) CreateWorkflow(ctx context.Context, w *datatypes.Workflow) erro
+
+func (c *client) ListWorkflowsBySIP(ctx context.Context, sipUUID uuid.UUID) ([]*datatypes.Workflow, error) {
+	dbw, err := c.ent.Workflow.Query().
+		WithSip().
```
Small improvement for the list, just fetch the SIP UUID `.WithSip(func(q *db.SIPQuery) { q.Select(sip.FieldUUID) })` (untested).

jraddaoui reviewed (changes requested):
SO nice to see this go, thanks @sevein!!

sevein committed:
Update: address feedback

sevein commented:
@jraddaoui I have addressed all your comments and observations. Please let me know if I missed anything.

These are the non-actionable comments shared in this pull request for future reference:
* https://github.com/artefactual-sdps/enduro/pull/1466#discussion_r2620315499
  » Return internal server error.
* https://github.com/artefactual-sdps/enduro/pull/1466#discussion_r2620325398
 » Replace SetStatus with SetSIPStatus or UpdateSIP.
* https://github.com/artefactual-sdps/enduro/pull/1466#discussion_r2620329180
  » Avoid using sql.NullTime in domain types.
* https://github.com/artefactual-sdps/enduro/pull/1466#discussion_r2620339213
  » Stop using "error ..." prefix in error messages
* https://github.com/artefactual-sdps/enduro/pull/1466#discussion_r2620349449
  » Adopt enums in workflow db type

jraddaoui reviewed (approved):
Thanks for the summary @sevein! LGTM!
